R/graded_condition.R

```diff
@@ -51,6 +65,67 @@ capture_errors <- function(expr, on_error = NULL) {
   )
 }
 
+
+sys_calls_most_helpful <- function() {
+  # borrowing ideas from rstudio/shiny/blob/2360bde1/R/conditions.R#L434-L448
+  calls <- sys.calls()
+  calls <- calls[-length(calls)] # not this function, obviously
+  callnames <- calls_name(calls)
+  
+  # We want to locate the last call in the stack that isn't just error handling
+  # infrastructure. The following functions are part of the stack when an error
+  # is captured and surfaced in gradethis, the hope is that the last call before
+  # the final block of these functions is reasonably informative.
+  hideable <- callnames %in% c(".handleSimpleError", "h", "on_error", "on_graded")
+  
+  # we also want to find the highest-level gradethis call, which probably
+  # provides the full context of the code generating the error
+  gradethis_pattern <- "^(gradethis:::?)?(grade_this|grade_result)"
+  idx_gradethis <- grep(gradethis_pattern, callnames)
+  
+  if (!length(idx_gradethis)) {
+    # next best guess, the function one level above the first function called
+    # in the function returned by grade_this()
+    idx_with_code_feedback <- which(callnames == "with_maybe_code_feedback")
+    if (length(idx_with_code_feedback)) {
+      idx_gradethis <- min(idx_with_code_feedback) - 1
+    }
+  }
+  
+  list(
+    first = if (length(idx_gradethis)) calls[[min(idx_gradethis)]],
+    last = calls[[max(which(!hideable))]]
+  )
+}
+
+calls_name <- function(calls) {
+  # borrowed from rstudio/shiny/blob/2360bde1/R/conditions.R#L64-L76
+  sapply(calls, function(call) {
+    if (is.function(call[[1]])) {
+      "<Anonymous>"
+    } else if (inherits(call[[1]], "call")) {
+      paste0(format(call[[1]]), collapse = " ")
+    } else if (typeof(call[[1]]) == "promise") {
+      "<Promise>"
+    } else {
+      paste0(as.character(call[[1]]), collapse = " ")
+    }
+  })
+}
+
+gradethis_fail_error_handler <- function(message, env = parent.frame(), ...) {
```

Comment:
Should we force these arguments?

```suggestion
gradethis_fail_error_handler <- function(message, env = parent.frame(), ...) {
  force(list(message, env, args = list(...)))
```

Reply:
yeah that's a good idea, thank you

Reply:
If there are other places in the code where args are not processed right away and the return value is used at a much later time, then it might be good to force those arguments as well.

(^^ update as you run across them.)